compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -244,14 +249,85 @@ private class JvmInlineClassLowering(private val context: JvmBackendContext) : F
             putValueArgument(0, argument)
         }
 
-    override fun visitCall(expression: IrCall): IrExpression {
-        val function = expression.symbol.owner
-        if (!function.isInlineClassFieldGetter)
-            return super.visitCall(expression)
-        val arg = expression.dispatchReceiver!!.transform(this, null)
-        return coerceInlineClasses(arg, function.dispatchReceiverParameter!!.type, expression.type)
+    private fun IrExpression.coerceToUnboxed() =
+        coerceInlineClasses(this, this.type, this.type.unboxInlineClass())
+
+    private fun IrClass.getEqualsImpl(): IrFunction {
+        val equals = functions.single { it.name.asString() == "equals" && it.overriddenSymbols.isNotEmpty() }
+        return manager.getReplacementFunction(equals)!!.function
+    }
+
+    private fun IrBuilderWithScope.specializeEqualsCall(left: IrExpression, right: IrExpression): IrExpression? {
+        // There's already special handling for null-comparisons in Equals.kt.
+        // We cannot specialize calls for which the first argument is already boxed.
+        if (left.isNullConst() || right.isNullConst() || left.type.unboxInlineClass() === left.type)
```

Comment:
We usually avoid using identity equality (`===`) unless it's the specifically required semantics. It's not a big deal here (and below), but it'd probably be more sensible to use `==` to avoid dependency on the fact that `unboxInlineClass` returns the original type in case the class is not `inline`.

Reply:
Fixed!